src/drivers/imx/esai.c

```diff
@@ -4,37 +4,456 @@
+{
+	direction = direction == SOF_IPC_STREAM_PLAYBACK;
+	/* FIFO enable */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFEN_MASK,
+			ESAI_xFCR_xFEN);
+
+	/* Write a few zero samples, one per channel, for initialization */
+	/* TODO determine the correct number of samples, assuming 2 */
+	if (direction) {
+		dai_write(dai, REG_ESAI_ETDR, 0);
+		dai_write(dai, REG_ESAI_ETDR, 0);
+	}
+
+	dai_update_bits(dai, REG_ESAI_xCR(direction),
+			direction ? ESAI_xCR_TE_MASK : ESAI_xCR_RE_MASK,
+			direction ? ESAI_xCR_TE(1) : ESAI_xCR_RE(1));
+
+	dai_update_bits(dai, REG_ESAI_xSMB(direction), ESAI_xSMB_xS_MASK,
+			ESAI_xSMB_xS(0x3));
+
+	dai_update_bits(dai, REG_ESAI_xSMA(direction), ESAI_xSMA_xS_MASK,
+			ESAI_xSMA_xS(0x3));
+
+	tracev_esai("ESAI_REGS_DUMP in esai_start");
+	esai_regs_dump(dai);
+}
+
+static void esai_stop(struct dai *dai, int direction)
+{
+	direction = direction == SOF_IPC_STREAM_PLAYBACK;
+
+	dai_update_bits(dai, REG_ESAI_xCR(direction),
+			direction ? ESAI_xCR_TE_MASK : ESAI_xCR_RE_MASK, 0);
+
+	dai_update_bits(dai, REG_ESAI_xSMA(direction), ESAI_xSMA_xS_MASK, 0);
+	dai_update_bits(dai, REG_ESAI_xSMB(direction), ESAI_xSMB_xS_MASK, 0);
+
+	/* disable and reset FIFO */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction),
+			ESAI_xFCR_xFR | ESAI_xFCR_xFEN, ESAI_xFCR_xFR);
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFR, 0);
+
+	tracev_esai("ESAI_REGS_DUMP in esai_stop");
+	esai_regs_dump(dai);
+}
+
 static int esai_trigger(struct dai *dai, int cmd, int direction)
 {
+	tracev_esai("ESAI: trigger");
+	switch (direction) {
+	case DAI_DIR_PLAYBACK:
+		tracev_esai("ESAI: playback trigger");
+		break;
+	case DAI_DIR_CAPTURE:
+		tracev_esai("ESAI: capture trigger");
+		break;
+	default:
+		trace_esai_error("ESAI: <INVALID DIRECTION %d> trigger", direction);
+	}
+
+	switch (cmd) {
+#define CASE(c) case COMP_ ## c: tracev_esai("ESAI: trigger cmd DAI_" #c ); break;
+#define CASE2(c, p) case COMP_ ## c: tracev_esai("ESAI: trigger cmd DAI_" #p); break;
+	case COMP_TRIGGER_START:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_START");
+		esai_start(dai, direction);
+		break;
+	case COMP_TRIGGER_STOP:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_STOP");
+		esai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_PAUSE:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_PAUSE_PUSH");
+		esai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_RELEASE:
+		tracev_esai("ESAI: trigger cmd DAI_TRIGGER_PAUSE_RELEASE");
+		esai_start(dai, direction);
+		break;
+	/* Remaining triggers are no-ops, just print */
+	CASE(TRIGGER_SUSPEND)
+	CASE(TRIGGER_RESUME)
+#undef CASE
+	default:
+		trace_esai_error("ESAI: invalid trigger cmd %d", cmd);
+		break;
+	}
 	return 0;
 }
 
 static int esai_probe(struct dai *dai)
 {
+	struct esai_pdata *pdata;
+
+	tracev_esai("ESAI: probe");
+	if (dai_get_drvdata(dai)) {
+		trace_esai_error("ESAI: Repeated probe, skipping");
+		return -EEXIST;
+	}
+	pdata = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
```

Comment:
```
#include <sof/lib/alloc.h>
#include <ipc/topology.h>
```

Reply:
I always tend to mess up the rule "from local to global", should <sof/*> come before <ipc/*> or the other way?